avro/src/main/java/com/fasterxml/jackson/dataformat/avro/AvroMapper.java

```diff
@@ -126,4 +146,85 @@ public AvroSchema schemaFrom(File schemaFile) throws IOException
         return new AvroSchema(new Schema.Parser().setValidate(true)
                 .parse(schemaFile));
     }
+
+    public static class AvroSchemaTypeResolverBuilder extends StdTypeResolverBuilder {
+
+        @Override
+        public TypeSerializer buildTypeSerializer(SerializationConfig config, JavaType baseType, Collection<NamedType> subtypes) {
+            // No default typing for Serialization - It's already encoded in the Schema object
+            return null;
+        }
+
+        @Override
+        public TypeDeserializer buildTypeDeserializer(DeserializationConfig config, JavaType baseType, Collection<NamedType> subtypes) {
+            if (subtypes == null) {
+                BeanDescription bean = config.introspectClassAnnotations(baseType.getRawClass());
+                AnnotatedClass  ac   = bean.getClassInfo();
+                subtypes = config.getSubtypeResolver().collectAndResolveSubtypesByTypeId(config, ac);
+            }
+            return new AvroSchemaTypeDeserializer(
+                baseType,
+                new AvroNativeTypeIdResolver(baseType, config.getTypeFactory(), subtypes),
+                getDefaultImpl() != null ? config.getTypeFactory().constructType(getDefaultImpl()) : baseType
+            );
+        }
+    }
+
+    public static class AvroSchemaTypeDeserializer extends TypeDeserializerBase {
+        protected AvroSchemaTypeDeserializer(JavaType baseType, TypeIdResolver idRes, JavaType defaultImpl) {
+            super(baseType, idRes, null, true, defaultImpl);
+        }
+
+        protected AvroSchemaTypeDeserializer(TypeDeserializerBase src, BeanProperty property) {
+            super(src, property);
+        }
+
+        @Override
+        public Object deserializeTypedFromObject(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromArray(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromScalar(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            return deserializeTypedFromAny(jsonParser, context);
+        }
+
+        @Override
+        public Object deserializeTypedFromAny(JsonParser jsonParser, DeserializationContext context) throws IOException {
+            // Primitives have no subtypes. Whatever the base type is, is correct!
+            if (!_baseType.isPrimitive() && jsonParser.canReadTypeId()) {
```

Comment:
The type deserializer gets called for the `byteValue` field (of type `byte`) in the `PrimitiveInteropTest`. Since in avro schema this is an "int with typeId java.lang.Byte" and Jackson barfs when it tries to process `java.lang.Byte` as a subtype of `byte`, this check seemed necessary. Let me know if there is a different way I should be handling this case.